core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleGetRecoveryCandidatesHandler.java

```diff
@@ -0,0 +1,187 @@
+
+            List<CloudInstance> instancesWithServicesNotRunning;
+
+            if (startedInstanceMetadataWithServicesNotRunning.isEmpty()) {
+                instancesWithServicesNotRunning = Collections.emptyList();
+            } else {
+                instancesWithServicesNotRunning = instanceMetaDataToCloudInstanceConverter.convert(startedInstanceMetadataWithServicesNotRunning,
+                        stack.getStack());
+                LOGGER.info("Collected {} downscale recovery candidates for hostGroup: {}", instancesWithServicesNotRunning.size(), request.getHostGroupName());
+            }
+
+            StopStartDownscaleGetRecoveryCandidatesResult result = new StopStartDownscaleGetRecoveryCandidatesResult(stack.getId(),
+                    instancesWithServicesNotRunning, request.getHostGroupName(), request.getHostIds());
+            notify(result, event);
+        } catch (Exception e) {
+            String errorMessage = String.format("Failed to collect downscale recovery candidates for stack: %s. " +
+                    "Proceeding with subsequent flow.", request.getResourceId());
+            LOGGER.warn(errorMessage, e);
+            StopStartDownscaleGetRecoveryCandidatesResult result = new StopStartDownscaleGetRecoveryCandidatesResult(request.getResourceId(),
+                    Collections.emptyList(), request.getHostGroupName(), request.getHostIds());
+            notify(result, event);
+        }
+    }
+
+    private List<CloudInstance> collectStartedInstancesFromCloudProvider(CloudConnector connector, AuthenticatedContext ac,
+            List<CloudInstance> cloudInstances) {
+        List<CloudVmInstanceStatus> vmInstanceStatuses = connector.instances().checkWithoutRetry(ac, cloudInstances);
+        return vmInstanceStatuses.stream()
+                .filter(vm -> InstanceStatus.STARTED.equals(vm.getStatus()))
+                .map(CloudVmInstanceStatus::getCloudInstance)
+                .collect(Collectors.toList());
+    }
+
+    //CHECKSTYLE:OFF
+    private List<InstanceMetadataView> collectRecoveryCandidates(List<InstanceMetadataView> startedInstances, DetailedHostStatuses detailedHostStatuses,
+            Optional<String> runtimeVersion) {
+        boolean servicesHealthCheckAllowed = CMRepositoryVersionUtil.isCmServicesHealthCheckAllowed(runtimeVersion);
+        Predicate<InstanceMetadataView> instanceHealthCheck =  im -> {
+            HostName hostName = hostName(im.getDiscoveryFQDN());
+            boolean hostUnhealthy = detailedHostStatuses.isHostUnHealthy(hostName);
+            boolean maintenanceMode = detailedHostStatuses.isHostInMaintenanceMode(hostName);
+            boolean hostDecommissioned = detailedHostStatuses.isHostDecommissioned(hostName);
+            boolean servicesNotRunning = detailedHostStatuses.areServicesNotRunning(hostName);
+            if (servicesHealthCheckAllowed) {
+                return (!hostUnhealthy && maintenanceMode && hostDecommissioned && servicesNotRunning)
```

Comment:
maintainanceMode maybe unnecessary here, because both value are permitted but the remaining booleans are the same

Reply:
It was decided to include these explicit boolean checks after observed behaviour with different clusters. The [document](https://docs.google.com/document/d/1CHO4kr1i5mwisPvbHGasaV6soTEOG1arujt49miRgFE/edit?usp=sharing) for the scenarios discusses this behaviour cc @sidseth

Reply:
The conditions are very explicit there to mirror the scenarios that we want to handle, which are detailed separately in a document. Yes, it is an unnecessary check - however it is far more maintainable going forward, once more scenarios are added to this.

Reply:
Sorry, but I don't understand this code how can more maintainable:
```if (servicesHealthCheckAllowed) {
    return (!hostUnhealthy && maintenanceMode && hostDecommissioned && servicesNotRunning) 
        || (!hostUnhealthy && !maintenanceMode && hostDecommissioned && servicesNotRunning);
} else {
    return (!hostUnhealthy && maintenanceMode && hostDecommissioned)
        || (!hostUnhealthy && !maintenanceMode && hostDecommissioned);
}
```
than this?
```
if (servicesHealthCheckAllowed) {
    return !hostUnhealthy && hostDecommissioned && servicesNotRunning;
} else {
    return !hostUnhealthy && hostDecommissioned;
}
```

Reply:
> which are detailed separately in a document
if I understand correctly, the doc says if maintenanceMode is enabled or not?

Reply:
Also, more conditions will be added to the document (and the code), as we find ones from which we think we can recover.

Reply:
Sorry, but I couldn't accept it. Next time will you create n^x branch to handle all of the `to be and not to be conditions` in the doc?

Reply:
also UT is missing for the whole class

Reply:
If usage of `maintenanceMode` will makes sense in a subsequent change, introduce it there, but based on the change of this PR, usage of `maintenanceMode` does not make sense and only increases the complexity of the code. If there will be more conditions later, refactor is absolute necessary, since the condition itself is hard to understand as of now too.

Reply:
Nobody will check or maintain the doc, but we need to understand this code piece a few months/years later. A more complex code is harder to understand later. Also, this was the reason why checkstyle error happened. Booleans complexity is increasing code complexity.

Reply:
This has been addressed

Reply:
looks good now

Reply:
looks good, thanks